File: 2306-create-binary-tree-from-descriptions/2306-create-binary-tree-from-descriptions.cpp
class Solution {
public:
    TreeNode* createBinaryTree(vector<vector<int>>& descriptions) {
        // Sets to track unique children and parents
        unordered_set<int> children, parents;
        // Map to store parent to children relationships
        unordered_map<int, vector<pair<int, int>>> parentToChildren;

        // Build graph from parent to child, and add nodes to HashSets
        for (auto& d : descriptions) {
            int parent = d[0], child = d[1], isLeft = d[2];
            parents.insert(parent);
            parents.insert(child);
            children.insert(child);
            parentToChildren[parent].emplace_back(child, isLeft);
        }

        // Find the root node by checking which node is in parents but not in children
        for (auto it = parents.begin(); it != parents.end();) {
            if (children.find(*it) != children.end()) {
                it = parents.erase(it);
            } else {
                ++it;
            }
        }
        TreeNode* root = new TreeNode(*parents.begin());

        // Starting from root, use BFS to construct binary tree
        queue<TreeNode*> queue;
        queue.push(root);

        while (!queue.empty()) {
            TreeNode* parent = queue.front();
            queue.pop();
            // Iterate over children of current parent
            for (auto& childInfo : parentToChildren[parent->val]) {
                int childValue = childInfo.first, isLeft = childInfo.second;
                TreeNode* child = new TreeNode(childValue);
                queue.push(child);
                // Attach child node to its parent based on isLeft flag
                if (isLeft == 1) {
                    parent->left = child;
                } else {
                    parent->right = child;
                }
            }
        }

        return root;
    }
};

File: 2306-create-binary-tree-from-descriptions/2306-create-binary-tree-from-descriptions.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def createBinaryTree(self, descriptions: List[List[int]]) -> Optional[TreeNode]:
        # Step 1: Organize data
        parent_to_children = {}
        all_nodes = set()
        children = set()
        
        for parent, child, is_left in descriptions:
            if parent not in parent_to_children:
                parent_to_children[parent] = []
            parent_to_children[parent].append((child, is_left))
            all_nodes.add(parent)
            all_nodes.add(child)
            children.add(child)
        
        # Step 2: Find the root
        root_val = (all_nodes - children).pop()
        
        # Step 3 & 4: Build the tree using DFS
        def dfs(val):
            node = TreeNode(val)
            if val in parent_to_children:
                for child, is_left in parent_to_children[val]:
                    if is_left:
                        node.left = dfs(child)
                    else:
                        node.right = dfs(child)
            return node
        
        return dfs(root_val)